vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -464,91 +409,132 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
 }
 
 facebook::jsi::Array ChakraRuntime::getPropertyNames(
-    const facebook::jsi::Object &obj) {
-  JsValueRef propertyNamesArrayRef;
-  checkException(JsGetOwnPropertyNames(objectRef(obj), &propertyNamesArrayRef));
-
-  JsPropertyIdRef propertyId;
-  checkException(JsGetPropertyIdFromName(L"length", &propertyId));
-  JsValueRef countRef;
-  checkException(JsGetProperty(propertyNamesArrayRef, propertyId, &countRef));
-  int count;
-  checkException(JsNumberToInt(countRef, &count));
-
-  auto result = createArray(count);
-  for (int i = 0; i < count; i++) {
-    JsValueRef index;
-    checkException(JsIntToNumber(i, &index));
-    JsValueRef propertyName;
-    checkException(
-        JsGetIndexedProperty(propertyNamesArrayRef, index, &propertyName));
-    result.setValueAtIndex(*this, i, createString(propertyName));
-  }
-
-  return result;
-}
+    const facebook::jsi::Object &object) {
+  constexpr const char *const jsGetPropertyNamesSource =
```

Comment:
>constexpr const char *const jsGetPropertyNamesSource = [](start = 2, length = 54)

This feels slightly strange; is there no easy way to keep the functionality in native code?

Reply:
I looked into the behavior of getPropertyNames for Hermes a while back, and I believe that it returns all enumerable properties of the Object, including the ones inherited through the prototype chain. To do this with Chakra(Core)'s native APIs, we would need to traverse the property chain (call JsGetPrototype repeatedly until we get undefined), and for each Object on the chain, call JsGetOwnPropertyNames and manually check whether each property is enumerable or not. The JS version is much easier to implement, since the for..in loop already loops over the properties we need to return.

---
In reply to: [336268580](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336268580) [](ancestors = 336268580)